### DISCUSSION TITLE
Add access modifier for namespaces

### DISCUSSION THREAD

**Discussion by captainkidd5** [dotnet/csharplang#4153] [Language Ideas] [answered]
For larger projects, it would be very helpful to create a class access modifier keyword such as "Closed" or something which would restrict the usage of a class to other classes in the same Namespace, so we don't have to create a separate DLL for smaller sections which really don't warrant it!

That way you can simply expose a single class to other namespaces, and keep the rest of the classes restricted to components of the exposed class :)

**Comment by CyrusNajmabadi** [comment, marked as answer]
You can do this with an analyzer.  In general, if you're asking for a feature to disallow or block off something currently legal, then we'll likely just push for this to be done with that mechanism.

**Reply by captainkidd5** [reply]
I'll have to look into analyzers a bit more, and how to customize them. That being said, I'm sort of amazed this feature hasn't been directly implemented into the language yet. It seems like an obvious encapsulation/organizational technique which prevents overcomplication and bloat with DLLs. Either way thank you for the response.

**Reply by HaloFour** [reply]
Accessibility modifiers are enforced by the runtime and the runtime really doesn't have a concept of "namespaces".  To the runtime the "namespace" is just the part of the class name.  There's no enforced encapsulation with namespaces like there is with Java 9+ Module System where two modules are not permitted to contribute classes to the same package/namespace.

**Reply by CyrusNajmabadi** [reply]
> t seems like an obvious encapsulation/organizational technique which prevents overcomplication 

I don't see it providing much benefit tbh.  Indeed, it would almost immediately run into problems.  For example, while i might see some scenarios where i would want a type filtered down to a namespace `Foo`, i would also want it available in `Foo.Bar`.  Also, namespaces are a cross-assembly construct.  So it would be just very weird.  You'd be filtering down types to a specific namespace so it would only usable there.  But anyone could still get access to it just by defining their own namespace with the same name.

As above though, when you want to restrict legal code, an analyzer is a simple and easy thing to do that already works *right now*.  So def the preferred way to go.
